File: 05_Simulation_ML/SmartSim/Polaris/Fortran/train_coDB/src/load_data.f
      module ssim

      use smartredis_client, only : client_type      

      type(client_type) :: client
      integer nnDB

      end module ssim

c ==================================================

      subroutine init_client(myrank)

      use iso_c_binding
      use ssim
      implicit none
      include "mpif.h"

      integer myrank, err

c     Initialize SmartRedis clients 
      if (myrank.eq.0) write(*,*) 'Initializing SmartRedis clients ... '
      if (nnDB.eq.1) then
         err = client%initialize(.false.) ! NOT using a clustered database (DB on 1 node only)
      else
         err = client%initialize(.true.) ! using a clustered database (DB on multiple nodes)
      endif
      if (err.ne.0) 
     &      write(*,*) "ERROR: client%initialize failed on rank ",myrank


      end subroutine init_client

c ==================================================

      program  data_loader

      use iso_c_binding
      use fortran_c_interop
      use ssim
      use, intrinsic :: iso_fortran_env
      implicit none
      include "mpif.h"

      real*8, allocatable, dimension (:,:) :: sendArr
      real*8 xmin, xmax, arrMLrun(2), x
      integer nSamples, nInputs, nOutputs, seed
      integer its, numts, i, err
      integer stepInfo(2), arrInfo(6)
      integer myrank, comm_size, ierr, tag, status(MPI_STATUS_SIZE), 
     &        nproc
      character*255 rank_key, sendArr_key
      character*255 ssdb

c     Initialize MPI
      call MPI_INIT(ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, comm_size, ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)
      nproc = comm_size
      call MPI_Barrier(MPI_COMM_WORLD,ierr)
      write(*,*) 'Hello from rank ',myrank
      flush(OUTPUT_UNIT)

c     Initialize SmartRedis clients
      nnDB = 1 
      call init_client(myrank)
      call MPI_Barrier(MPI_COMM_WORLD,ierr)
      if (myrank.eq.0) write(*,*) 'All SmartRedis clients initialized'

c     Write the DB IP address to file
      call get_environment_variable("SSDB", ssdb)
      if (myrank.eq.0) then
         open (unit=25, file='SSDB.dat', status='replace')
         write(25,'(a)',advance='no') trim(adjustl(ssdb))
         close(25)
      endif

c     Set parameters for array of random numbers to be set as training data
c     In this example we create training data for a simple function
c     y=f(x), which has 1 input (x) and 1 output (y)
c     The domain for the function is from 0 to 10
c     Tha training data is obtained from a uniform distribution over the domain
      nSamples = 64
      nInputs = 1
      nOutputs = 1
      allocate(sendArr(nSamples,nInputs+nOutputs))
      seed = myrank+1
      call RANDOM_SEED(seed)
      xmin = 0.0
      xmax = 10.0


c     Send array used to communicate whether to keep running data loader or ML
      if (myrank.eq.0) then
         arrMLrun = 1.0
         err = client%put_tensor("check-run", arrMLrun, shape(arrMLrun))
         if (err.ne.0) write(*,*)
     &            "ERROR: client%put_tensor failed on rank ",myrank
      endif


c     Send some information regarding the training data size
      if (myrank.eq.0) then
         arrInfo(1) = nSamples
         arrInfo(2) = nInputs+nOutputs
         arrInfo(3) = nInputs
         arrInfo(4) = nproc
         arrInfo(5) = nproc
         arrInfo(6) = myrank
         err = client%put_tensor("sizeInfo", arrInfo, shape(arrInfo))
         if (err.ne.0) then 
            write(*,*) "ERROR: client%put_tensor failed on rank ",myrank
         else
            write(*,*) "Sent size info of training data to database"
         endif
      endif


c     Generate first part of the key for the training data
c     The key will be tagged with the rank ID and the time step number
      rank_key = "y."
      if (myrank.lt.10) then
         write (rank_key, "(A2,I1)") trim(rank_key), myrank
      elseif (myrank.lt.100) then
         write (rank_key, "(A2,I2)") trim(rank_key), myrank
      elseif (myrank.lt.1000) then
         write (rank_key, "(A2,I3)") trim(rank_key), myrank
      endif


c     Emulate integration of PDEs with a do loop
      numts = 1000
      do its=1,numts
         ! sleep for a few seconds to emulate the time required by PDE integration
         call sleep (10)

         ! first off check if ML is done training, if so exit from loop
         err = client%unpack_tensor("check-run", arrMLrun, 
     &                              shape(arrMLrun))
         if (err.ne.0) write(*,*) 
     &           "ERROR: client%unpack_tensor failed on rank ",myrank
         if (arrMLrun(1).lt.0.5) exit

         ! generate the training data for the polynomial y=f(x)=x**2 + 3*x + 1
         ! place output in first column, input in second column
         do i=1,nSamples
            call RANDOM_NUMBER(x)
            x = xmin + (xmax-xmin)*x
            sendArr(i,2) = x
            sendArr(i,1) = x**2 + 3*x +1
         enddo

         ! Append the ts number to the key so data doesn't get overwritten in database
         if (myrank.lt.10) then
            write (sendArr_key, "(A3,A1,I0)") trim(rank_key),'.',its
         elseif (myrank.lt.100) then
            write (sendArr_key, "(A4,A1,I0)") trim(rank_key),'.',its
         elseif (myrank.lt.1000) then
            write (sendArr_key, "(A5,A1,I0)") trim(rank_key),'.',its
         endif
         ! send training data to database
         if (myrank.eq.0) write(*,*) 
     &            'Sending training data to database with key ',
     &            trim(sendArr_key), ' and shape ',
     &            shape(sendArr)
         err = client%put_tensor(trim(sendArr_key), sendArr, 
     &                           shape(sendArr))
         call MPI_Barrier(MPI_COMM_WORLD,ierr)
         if (err.ne.0) then 
            write(*,*) "ERROR: client%put_tensor failed on rank ",myrank
         else
            if (myrank.eq.0) write(*,*) 
     &                'All ranks finished sending training data'
         endif
         ! Send also the time step number, used by ML program to determine 
         ! when new training data is available
         if (myrank.eq.0) then
            stepInfo(1) = its
            stepInfo(2) = 0.0
            err = client%put_tensor("step", stepInfo, shape(stepInfo))
            if (err.ne.0) write(*,*)
     &             "ERROR: client%put_tensor failed on rank ",myrank
         endif
      enddo
c     Finilization stuff
      if (myrank.eq.0) write(*,*) "Exiting ... "
      deallocate(sendArr)
      call MPI_FINALIZE(ierr)
      end program data_loader

File: couplingSimulationML/SmartSim/Polaris/Python/train_clDB/src/load_data.py
import sys
import os
import argparse
from time import sleep, perf_counter
import numpy as np
import logging
from smartredis import Client

def setup_logger(name, log_file, level=logging.INFO):
    """To setup as many loggers as you want"""
    handler = logging.FileHandler(log_file,mode='w')
    formatter = logging.Formatter('%(message)s')
    handler.setFormatter(formatter)
    logger = logging.getLogger(name)
    logger.setLevel(level)
    logger.addHandler(handler)
    return logger

def init_client(args, logger_init):
    if (args.dbnodes==1):
        tic = perf_counter()
        client = Client(cluster=False)
        toc = perf_counter()
    else:
        tic = perf_counter()
        client = Client(cluster=True)
        toc = perf_counter()
    if (args.logging=='verbose'):
        logger_init.info('%.8e',toc-tic)
    return client

def main():
    # Import and initialize MPI
    from mpi4py import MPI
    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()
    name = MPI.Get_processor_name()
    print(f'Rank {rank}/{size} says hello from node {name}') 
    comm.Barrier()
    sys.stdout.flush()

    # Parse arguments
    parser = argparse.ArgumentParser(description='')
    parser.add_argument('--dbnodes',default=1,type=int,help='Number of database nodes')
    parser.add_argument('--ppn',default=4,type=int,help='Number of processes per node')
    parser.add_argument('--logging',default='no',help='Level of performance logging')
    args = parser.parse_args()
    if (rank==0):
        print(f'\nRunning with {args.dbnodes} DB nodes')
        print(f'and with {args.ppn} processes per node \n')
        sys.stdout.flush()

    # Create log files
    time_meta = 0.
    if (args.logging=='verbose'):
        logger_init = setup_logger('client_init', f'client_init_{rank}.log')
        logger_meta = setup_logger('meta', f'meta_data_{rank}.log')
        logger_data = setup_logger('train_data', f'train_data_{rank}.log')
    else:
        logger_init = None

    # Initialize SmartRedis clients
    client = init_client(args, logger_init)
    comm.Barrier()
    if (rank==0):
        print('All SmartRedis clients initialized')
        sys.stdout.flush()
    
    # Set parameters for array of random numbers to be set as inference data
    # In this example we create inference data for a simple function
    # y=f(x), which has 1 input (x) and 1 output (y)
    # The domain for the function is from 0 to 10
    # The inference data is obtained from a uniform distribution over the domain
    nSamples = 64
    xmin = 0.0 
    xmax = 10.0
    nInputs = 1
    nOutputs = 1

    # Send array used to communicate whether to keep running data loader or ML
    if (rank == 0):
        arrMLrun = np.array([1, 1])
        tic = perf_counter()
        client.put_tensor('check-run', arrMLrun)
        toc = perf_counter()
        time_meta = time_meta + (toc - tic)

    # Send some information regarding the training data size
    if (rank == 0):
        arrInfo = np.array([nSamples, nInputs+nOutputs, nInputs,
                            size, args.ppn, rank])
        tic = perf_counter()
        client.put_tensor('sizeInfo', arrInfo)
        toc = perf_counter()
        time_meta = time_meta + (toc - tic)
        print('Sent size info of training data to database')

    # Emulate integration of PDEs with a do loop
    numts = 1000
    stepInfo = np.zeros(2, dtype=int)
    for its in range(numts):
        # Sleep for a few seconds to emulate the time required by PDE integration
        sleep(2)

        # First off check if ML is done training, if so exit from loop
        tic = perf_counter()
        arrMLrun = client.get_tensor('check-run')
        toc = perf_counter()
        time_meta = time_meta + (toc - tic)
        if (arrMLrun[0]<0.5):
            break

        # Generate the training data for the polynomial y=f(x)=x**2 + 3*x + 1
        # place output in first column and input in second column
        inputs = np.random.uniform(low=xmin, high=xmax, size=(nSamples,1))
        outputs = inputs**2 + 3*inputs + 1
        sendArr = np.concatenate((outputs, inputs), axis=1)

        # Send training data to database
        send_key = 'y.'+str(rank)+'.'+str(its+1)
        if (rank==0):
            print(f'Sending training data with key {send_key} and shape {sendArr.shape}')
        tic = perf_counter()
        client.put_tensor(send_key, sendArr)
        toc = perf_counter()
        if (args.logging=='verbose'):
            logger_data.info('%.8e',toc-tic)
        comm.Barrier()
        if (rank==0):
            print(f'All ranks finished sending training data')

        # Send the time step number, used by ML program to determine
        # when new data is available
        if (rank == 0):
            stepInfo[0] = int(its+1)
            tic = perf_counter()
            client.put_tensor('step', stepInfo)
            toc = perf_counter()
            time_meta = time_meta + (toc - tic)

        sys.stdout.flush()

    if (args.logging=='verbose'):
        logger_meta.info('%.8e',time_meta)

    if (rank==0):
        print('Exiting ...')


if __name__ == '__main__':
    main()
